privacyidea/api/lib/prepolicy.py

```diff
@@ -182,8 +230,23 @@ def set_random_pin(request=None, action=None):
         raise TokenAdminError("You need to specify a policy '{0!s}' in scope "
                               "{1!s}.".format(ACTION.OTPPINSETRANDOM, role))
     elif len(pin_pols) == 1:
-        log.debug("Creating random OTP PIN with length {0!s}".format(list(pin_pols)[0]))
-        request.all_data["pin"] = generate_password(size=int(list(pin_pols)[0]))
+        # check pin contents policy per token type, otherwise fall back
+        tokentype = request.all_data.get("type", "hotp")
```

Comment:
The "set_random_pin" request can be called by a button in the UI as a special request for one token. I think there is token tokentype in this request. So we can get it this way:

~~~~python
tokentype = get_token_type(request.all_data.get("serial"))
~~~~

get_token_type is imported from privacyidea.lib.token

Reply:
done